### PR TITLE
Amazon Alexa extension for Quarkus Native integration with AWS Lambda

### PR DESCRIPTION
**Pull request by oztimpower**
**Summary:**

Introduce a new extension, to assist the native image compilation of Amazon Alexa AWS Lambda functions.  Alexa can also be hosted by any HTTPS endpoint.

The Amazon Alexa ASK Java SDK v2 uses the RequestStreamHandler interface, via subclassing the SkillStreamHandler classes.  With the merge of commits to enable Amazon Alexa to be integrated with Quarkus, the remaining piece was support for Quarkus native.

Commits that added the functionality required for Alexa were #7866, #7985 and documentation #7927

The AWS ASK SDK v2 uses a builder pattern for the Jackson deserialisation, which needs to be added as reflective classes to the native build.  I've created a simple scanner to do this dynamically via identifying all classes annotated with JsonDeserialize.class. As common with all AWS clients, a dynamic proxy is required also.  It is quite tedious to do all this manually if a Quarkus extension was not used.

FYI, I'll update documentation in a subsequent PR if this one is accepted.

**Tested by**:

**Install new extension:**
1. git clone -b ext/alexa https://github.com/oztimpower/quarkus.git
2. cd quarkus
3. cd extensions/amazon-alexa
4. mvn install

**Alexa Test Project:**
1. git clone https://github.com/oztimpower/alexa-quarkus.git
2. cd alexa-quarkus
3. mvn -Pnative package -Dquarkus.native.container-build=true

The output of the native image will show the number of classes included for build time.

`Printing 20792 classes of type BUILD_TIME to /project/reports/build_time_classes_20200330_052538.txt
`

4. sam local invoke --template sam.native.yaml --event payload.json

A positive outcome is expected, where Quarkus Native is able to produce a valid response to the Lambda test payload.json.

```
END RequestId: d9815e18-3c7c-1881-120e-af6ebed0bb81
REPORT RequestId: d9815e18-3c7c-1881-120e-af6ebed0bb81	Init Duration: 235.37 ms	Duration: 103.46 ms	Billed Duration: 200 ms	Memory Size: 256 MB	Max Memory Used: 51 MB	

{"version":"1.0","userAgent":"ask-java/2.28.0 Java/11.0.6 templateResolver","response":{"outputSpeech":{"type":"SSML","ssml":"\u003cspeak\u003eHi \u003calexa:name type=\"first\" personId=\"amzn1.ask.account.AA274FLEVWIWU6CTPZQ25XW6P7XGAJJ4YB7OPEFKULHJRUQDIO5AFS2UNPLBPG6BZKTGRTH6BDQZJ3SNAC6ELTPWT4DALVM5WZGC59JHXBZQRZ4V646JEF2EKO3DVMWW65E67NPBTLIGV4HUDK45JFAQWAY5C7OGWQNIOLWNS4M6WNLIMQWU5HUEHDXAT7QTONKHS7ILR2V5EOY\"/\u003e\u003c/speak\u003e"}}}

```

**Negative Test:**

1. Comment out the new extension in the project POM

```
<!--
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-amazon-alexa</artifactId>
            <version>${quarkus.version}</version>
        </dependency>
-->

```

2. Recompile
mvn  -Pnative package -Dquarkus.native.container-build=true

The number of classes included at build time drops significantly.

`Printing 11212 classes of type BUILD_TIME to /project/reports/build_time_classes_20200330_055547.txt
`

3. sam local invoke --template sam.native.yaml --event payload.json

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Builder class com.amazon.ask.model.RequestEnvelope$Builder does not have build method (name: 'build')

END RequestId: 411c18ec-af00-1623-1ff3-795de76f32ed
REPORT RequestId: 411c18ec-af00-1623-1ff3-795de76f32ed	Init Duration: 143.56 ms	Duration: 49.91 ms	Billed Duration: 100 ms	Memory Size: 256 MB	Max Memory Used: 35 MB	

{"errorType":"com.amazon.ask.exception.AskSdkException","errorMessage":"Deserialization error"}

```

/cc @patriot1burke 
/cc @gsmet 
/cc @geoand

### CODE REVIEW

File: extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java
Comment (geoand) [inline extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Can we try and use Jandex to do this? 
That is pretty much what all the other extensions do.

Comment (gsmet) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Also, I think we should do that in the Jackson extension and just have this one depend on the Jackson extension.

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (gsmet) [inline extensions/amazon-lambda-alexa/runtime/pom.xml]: Is it specific to Amazon Lambda or could be used to use Alexa in another context?

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (gsmet) [inline extensions/amazon-lambda-alexa/runtime/pom.xml]: I'm surprised we don't have a dependency to some Alexa SDK?

File: extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Yes, I was thinking it would be nice to have standard functionality that could automatically scan the class path and identify the JsonDeserialize usage.  The one caveat to this is that the Alexa ASK SDK also has a static Builder inner class, that requires ClassName$Builder to be added.

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: If I understand the question correctly, no Amazon Alexa would only be used with AWS Lambda in this context.  Alexa is driven by custom AWS Lambda code "Skills" that give it functionality (for voice, video et al).

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: The Alexa SDK dependency is in the user project, since they need to subclass the SkillStreamHandler, and leverage other SDK dependencies to use Alexa with AWS Lambda.  Hence no need to include in the Quarkus dependencies.

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: Oh I should add yes, you can also write Alexa Skills outside of AWS Lambda, any HTTPS interface actually, so containers would be applicable also, I guess that's what you were asking -- in terms of how to name the extension.

File: extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Interestingly the Jackson extension potentially does have code to handle this & the builder already, yet it doesn't seem to pickup the JsonDeserialize annotation.

The combinedBuildIndex is small in size.  

```
  // handle the various @JsonDeserialize cases
        for (AnnotationInstance deserializeInstance : index.getAnnotations(JSON_DESERIALIZE)) {
            AnnotationTarget annotationTarget = deserializeInstance.target();
            if (CLASS.equals(annotationTarget.kind())) {
                DotName dotName = annotationTarget.asClass().name();
                if (!ignoredDotNames.contains(dotName)) {
                    addReflectiveHierarchyClass(dotName);
                }

                AnnotationValue annotationValue = deserializeInstance.value("builder");
                if (null != annotationValue && AnnotationValue.Kind.CLASS.equals(annotationValue.kind())) {
                    DotName builderClassName = annotationValue.asClass().name();
                    if (!BUILDER_VOID.equals(builderClassName)) {
                        addReflectiveHierarchyClass(builderClassName);
                    }
                }
            }
    } 
```

Comment (geoand) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: You might need to create an `IndexDependencyBuildItem` in order to get that dependency into the inde (if it's not already in).

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (gsmet) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: Yeah, what I was wondering is if this could be a general purpose Alexa extension that we could use for all usage.

That would mean a few things:
- it wouldn't depend on Amazon Lambda. Does it right now? I saw the dependency but I'm not sure it's strictly needed.
- we would add the Alexa dependency to the runtime artifact so that just adding it drags the Alexa artifact.
- obviously, we would change the name to be more generic :).

Comment (gsmet) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: Note that these are suggestions. I'm not accustomed to the Amazon ecosystem so feel free to push back if these are bad ideas.

File: extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Got this working with Jandex, using the IndexDependencyBuildItem as a BuildStep.

Code is quite simple now, all it does is draw in the Alexa dependencies, and then it uses the Jackson extension to do the work of identifying the JsonDeserialize & Builder classes.

Thanks to @stuartwdouglas for helping me today.

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: Renamed it to amazon-alexa, makes sense, no need to couple it with the Amazon Lambda extension.

File: extensions/amazon-lambda-alexa/runtime/pom.xml
Comment (oztimpower) [reply extensions/amazon-lambda-alexa/runtime/pom.xml]: Added the dependency to the runtime BOM and also the runtime pom.  It needs to exclude logging, and an internal conflict.  User project doesn't require the explicit dependency now.

File: extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java
Comment (geoand) [reply extensions/amazon-lambda-alexa/deployment/src/main/java/io/quarkus/amazon/lambda/alexa/AlexaProcessor.java]: Excellent!

**Review by geoand — APPROVED** [review]

**Comment by oztimpower** [comment]
@geoand , thanks for approving.  Any chance you could restart the 1 failed process which seemed to timeout?  CI/TCKS Test failed (no idea what that is!).

I'm presenting an Online Summit for Amazon, titled Supersonic Serverless Java on AWS Lambda, and using Alexa as my Quarkus native demo, to show the cold start difference; it would be great if this is in the code since I'm promoting it.  Alexa is a complex use case for AWS Lambda, with this working, it is another step forward for the maturity of Quarkus (and moving from Preview to GA) for AWS Lambda support.

**Comment by geoand** [comment]
@oztimpower sounds cool!

I restarted all tests (in GH Pipelines we can't just restart one unfortunately...)

**Comment by oztimpower** [comment]
@geoand Your Devoxx presentation last year on Quarkus was pretty impressive I have to add, and the demos with Spring. I've watched it and refer it to people when I talk about Quarkus to AWS customers.

**Comment by geoand** [comment]
Thank you very much for your kind words @oztimpower :)

**Comment by geoand** [comment]
I know @iocanel will really really love this work! He can't get enough of tinkering with Alexa :wink:

**Comment by geoand** [comment]
Let's also wait for an approval from the master, @gsmet :)

**Pull request merged**
